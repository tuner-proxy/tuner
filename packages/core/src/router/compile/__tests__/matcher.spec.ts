import {
  MatchInfo,
  createHostnameMatcher,
  createPathnameMatcher,
  createPortMatcher,
  createProtocolMatcher,
  createSearchParamsMatcher,
  createURLMatcher,
} from '../matcher';

describe('matcher', () => {
  match('protocol', createProtocolMatcher, [
    {
      pattern: '',
      toMatch: createCases('protocol', ['http:']),
    },
    {
      pattern: 'http:',
      toMatch: createCases('protocol', ['http:']),
      toNotMatch: createCases('protocol', ['https:']),
    },
    {
      pattern: 'http|https:',
      toMatch: createCases('protocol', ['http:', 'https:']),
      toNotMatch: createCases('protocol', ['ws:', 'wss:']),
    },
  ]);

  match('hostname', createHostnameMatcher, [
    {
      pattern: '',
      toMatch: createCases('hostname', ['example.com']),
    },
    {
      pattern: 'example.com',
      toMatch: createCases('hostname', ['example.com']),
      toNotMatch: createCases('hostname', ['foo.com', 'foo.example.com']),
    },
    {
      pattern: '*.example.com',
      toMatch: createCases('hostname', ['foo.example.com']),
      toNotMatch: createCases('hostname', [
        'example.com',
        'foo.com',
        'foo.bar.example.com',
      ]),
    },
    {
      pattern: '*.*.example.com',
      toMatch: createCases('hostname', [
        'foo.example.com',
        'foo.bar.example.com',
        'foo.bar.baz.example.com',
      ]),
      toNotMatch: createCases('hostname', [
        'example.com',
        'foo.com',
        'foo.example.com.net',
      ]),
    },
    {
      pattern: '[1.0.0.0/16]',
      toMatch: createCases('hostname', ['1.0.0.0', '1.0.0.1', '1.0.1.1']),
      toNotMatch: createCases('hostname', [
        '1.1.1.1',
        '2.0.0.0',
        'example.com',
      ]),
    },
    {
      pattern: '[192.168.1.128-192.168.1.255]',
      toMatch: createCases('hostname', ['192.168.1.128', '192.168.1.255']),
      toNotMatch: createCases('hostname', [
        '192.168.1.1',
        '192.168.2.0',
        'example.com',
      ]),
    },
    {
      pattern: '[2001::abc-2001::1:ffff]',
      toMatch: createCases('hostname', ['2001::bbb', '2001::1:1']),
      toNotMatch: createCases('hostname', [
        '0.0.0.0',
        '2001::2:1',
        'example.com',
      ]),
    },
  ]);

  match('port', createPortMatcher, [
    {
      pattern: '',
      toMatch: createCases('port', [80, '80']),
    },
    {
      pattern: '80',
      toMatch: createCases('port', [80, '80']),
      toNotMatch: createCases('port', [443, '443']),
    },
    {
      pattern: '80|443',
      toMatch: createCases('port', [80, '80', 443, '443']),
      toNotMatch: createCases('port', [21, '21']),
    },
  ]);

  match('search', createSearchParamsMatcher, [
    {
      pattern: '',
      toMatch: createCases('searchParams', [
        new URLSearchParams(''),
        new URLSearchParams('?foo=bar&baz'),
      ]),
    },
    {
      pattern: '?foo',
      toMatch: createCases('searchParams', [
        new URLSearchParams('?foo'),
        new URLSearchParams('?foo=bar'),
        new URLSearchParams('?foo&baz'),
        new URLSearchParams('?foo=bar&baz'),
      ]),
      toNotMatch: createCases('searchParams', [
        new URLSearchParams(''),
        new URLSearchParams('?bar'),
      ]),
    },
    {
      pattern: '?foo=bar',
      toMatch: createCases('searchParams', [
        new URLSearchParams('?foo=bar'),
        new URLSearchParams('?foo=bar&baz'),
      ]),
      toNotMatch: createCases('searchParams', [
        new URLSearchParams(''),
        new URLSearchParams('?foo'),
        new URLSearchParams('?foo=baz'),
      ]),
    },
  ]);

  match('pathname', createPathnameMatcher, [
    {
      pattern: '',
      toMatch: createCases('pathname', ['/', '/foo/bar']),
    },
    {
      pattern: '/absolute/path-name',
      toMatch: createCases('pathname', ['/absolute/path-name']),
      toNotMatch: createCases('pathname', [
        '/absolute',
        '/absolute/path-name/',
        '/absolute/path-name/child',
      ]),
    },
    {
      pattern: '/with/:param',
      toMatch: createCases('pathname', ['/with/param-value']),
      toNotMatch: createCases('pathname', [
        '/with',
        '/with/',
        '/with/param/child',
      ]),
    },
    {
      pattern: '/with/::longParam',
      toMatch: createCases('pathname', [
        '/with/param-value',
        '/with/long/param/value',
      ]),
      toNotMatch: createCases('pathname', ['/with', '/with/', '/not-match/']),
    },
  ]);

  match('full', createURLMatcher, [
    {
      pattern:
        'http|https://*.*.example.com:80|443/pathname/::param?foo&bar=baz',
      toMatch: [
        {
          protocol: 'http:',
          hostname: 'foo.example.com',
          port: 443,
          pathname: '/pathname/with/param',
          searchParams: new URLSearchParams('?foo=baz&bar=baz'),
        },
      ],
      toNotMatch: [
        {
          protocol: 'ws:',
        },
        {
          protocol: 'http:',
          port: 21,
        },
        {
          protocol: 'http:',
          port: 443,
          hostname: 'foo.com',
        },
        {
          protocol: 'http:',
          port: 443,
          hostname: 'foo.example.com',
          searchParams: new URLSearchParams('?foo&bar=bak'),
        },
      ],
    },
  ]);
});

interface TestCase {
  pattern: string;
  toMatch?: Partial<MatchInfo>[];
  toNotMatch?: Partial<MatchInfo>[];
}

function match(
  name: string,
  createMatcher: (pattern: string) => (info: any) => unknown,
  caseList: TestCase[],
) {
  for (const { pattern, toMatch, toNotMatch } of caseList) {
    for (const info of toMatch || []) {
      it(`${name} '${pattern}' match ${stringify(info)}`, () => {
        const res = createMatcher(pattern)(info);
        expect(res).toBeTruthy();
        expect(res).toMatchSnapshot();
      });
    }
    for (const info of toNotMatch || []) {
      it(`${name} '${pattern}' not match ${stringify(info)}`, () => {
        const res = createMatcher(pattern)(info);
        expect(res).toBeFalsy();
        expect(res).toMatchSnapshot();
      });
    }
  }
}

function stringify(value: any) {
  return JSON.stringify(value, (key, value) => {
    if (value instanceof URLSearchParams) {
      return `${value}`;
    }
    return value;
  });
}

function createCases(key: keyof MatchInfo, values: unknown[]) {
  return values.map((value) => ({ [key]: value }));
}
