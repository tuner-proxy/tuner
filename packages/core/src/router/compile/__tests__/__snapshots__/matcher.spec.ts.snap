// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matcher full 'http|https://*.*.example.com:80|443/pathname/::param?foo&bar=baz' match {"protocol":"http:","hostname":"foo.example.com","port":443,"pathname":"/pathname/with/param","searchParams":"foo=baz&bar=baz"} 1`] = `
{
  "param": "with/param",
}
`;

exports[`matcher full 'http|https://*.*.example.com:80|443/pathname/::param?foo&bar=baz' not match {"protocol":"http:","port":21} 1`] = `undefined`;

exports[`matcher full 'http|https://*.*.example.com:80|443/pathname/::param?foo&bar=baz' not match {"protocol":"http:","port":443,"hostname":"foo.com"} 1`] = `undefined`;

exports[`matcher full 'http|https://*.*.example.com:80|443/pathname/::param?foo&bar=baz' not match {"protocol":"http:","port":443,"hostname":"foo.example.com","searchParams":"foo=&bar=bak"} 1`] = `undefined`;

exports[`matcher full 'http|https://*.*.example.com:80|443/pathname/::param?foo&bar=baz' not match {"protocol":"ws:"} 1`] = `undefined`;

exports[`matcher hostname '' match {"hostname":"example.com"} 1`] = `true`;

exports[`matcher hostname '*.*.example.com' match {"hostname":"foo.bar.baz.example.com"} 1`] = `true`;

exports[`matcher hostname '*.*.example.com' match {"hostname":"foo.bar.example.com"} 1`] = `true`;

exports[`matcher hostname '*.*.example.com' match {"hostname":"foo.example.com"} 1`] = `true`;

exports[`matcher hostname '*.*.example.com' not match {"hostname":"example.com"} 1`] = `false`;

exports[`matcher hostname '*.*.example.com' not match {"hostname":"foo.com"} 1`] = `false`;

exports[`matcher hostname '*.*.example.com' not match {"hostname":"foo.example.com.net"} 1`] = `false`;

exports[`matcher hostname '*.example.com' match {"hostname":"foo.example.com"} 1`] = `true`;

exports[`matcher hostname '*.example.com' not match {"hostname":"example.com"} 1`] = `false`;

exports[`matcher hostname '*.example.com' not match {"hostname":"foo.bar.example.com"} 1`] = `false`;

exports[`matcher hostname '*.example.com' not match {"hostname":"foo.com"} 1`] = `false`;

exports[`matcher hostname 'example.com' match {"hostname":"example.com"} 1`] = `true`;

exports[`matcher hostname 'example.com' not match {"hostname":"foo.com"} 1`] = `false`;

exports[`matcher hostname 'example.com' not match {"hostname":"foo.example.com"} 1`] = `false`;

exports[`matcher pathname '' match {"pathname":"/"} 1`] = `{}`;

exports[`matcher pathname '' match {"pathname":"/foo/bar"} 1`] = `{}`;

exports[`matcher pathname '/absolute/path-name' match {"pathname":"/absolute/path-name"} 1`] = `{}`;

exports[`matcher pathname '/absolute/path-name' not match {"pathname":"/absolute"} 1`] = `undefined`;

exports[`matcher pathname '/absolute/path-name' not match {"pathname":"/absolute/path-name/"} 1`] = `undefined`;

exports[`matcher pathname '/absolute/path-name' not match {"pathname":"/absolute/path-name/child"} 1`] = `undefined`;

exports[`matcher pathname '/with/::longParam' match {"pathname":"/with/long/param/value"} 1`] = `
{
  "longParam": "long/param/value",
}
`;

exports[`matcher pathname '/with/::longParam' match {"pathname":"/with/param-value"} 1`] = `
{
  "longParam": "param-value",
}
`;

exports[`matcher pathname '/with/::longParam' not match {"pathname":"/not-match/"} 1`] = `undefined`;

exports[`matcher pathname '/with/::longParam' not match {"pathname":"/with"} 1`] = `undefined`;

exports[`matcher pathname '/with/::longParam' not match {"pathname":"/with/"} 1`] = `undefined`;

exports[`matcher pathname '/with/:param' match {"pathname":"/with/param-value"} 1`] = `
{
  "param": "param-value",
}
`;

exports[`matcher pathname '/with/:param' not match {"pathname":"/with"} 1`] = `undefined`;

exports[`matcher pathname '/with/:param' not match {"pathname":"/with/"} 1`] = `undefined`;

exports[`matcher pathname '/with/:param' not match {"pathname":"/with/param/child"} 1`] = `undefined`;

exports[`matcher port '' match {"port":"80"} 1`] = `true`;

exports[`matcher port '' match {"port":80} 1`] = `true`;

exports[`matcher port '80' match {"port":"80"} 1`] = `true`;

exports[`matcher port '80' match {"port":80} 1`] = `true`;

exports[`matcher port '80' not match {"port":"443"} 1`] = `false`;

exports[`matcher port '80' not match {"port":443} 1`] = `false`;

exports[`matcher port '80|443' match {"port":"80"} 1`] = `true`;

exports[`matcher port '80|443' match {"port":"443"} 1`] = `true`;

exports[`matcher port '80|443' match {"port":80} 1`] = `true`;

exports[`matcher port '80|443' match {"port":443} 1`] = `true`;

exports[`matcher port '80|443' not match {"port":"21"} 1`] = `false`;

exports[`matcher port '80|443' not match {"port":21} 1`] = `false`;

exports[`matcher protocol '' match {"protocol":"http:"} 1`] = `true`;

exports[`matcher protocol 'http:' match {"protocol":"http:"} 1`] = `true`;

exports[`matcher protocol 'http:' not match {"protocol":"https:"} 1`] = `false`;

exports[`matcher protocol 'http|https:' match {"protocol":"http:"} 1`] = `true`;

exports[`matcher protocol 'http|https:' match {"protocol":"https:"} 1`] = `true`;

exports[`matcher protocol 'http|https:' not match {"protocol":"ws:"} 1`] = `false`;

exports[`matcher protocol 'http|https:' not match {"protocol":"wss:"} 1`] = `false`;

exports[`matcher search '' match {"searchParams":""} 1`] = `true`;

exports[`matcher search '' match {"searchParams":"foo=bar&baz="} 1`] = `true`;

exports[`matcher search '?foo' match {"searchParams":"foo="} 1`] = `true`;

exports[`matcher search '?foo' match {"searchParams":"foo=&baz="} 1`] = `true`;

exports[`matcher search '?foo' match {"searchParams":"foo=bar"} 1`] = `true`;

exports[`matcher search '?foo' match {"searchParams":"foo=bar&baz="} 1`] = `true`;

exports[`matcher search '?foo' not match {"searchParams":""} 1`] = `false`;

exports[`matcher search '?foo' not match {"searchParams":"bar="} 1`] = `false`;

exports[`matcher search '?foo=bar' match {"searchParams":"foo=bar"} 1`] = `true`;

exports[`matcher search '?foo=bar' match {"searchParams":"foo=bar&baz="} 1`] = `true`;

exports[`matcher search '?foo=bar' not match {"searchParams":""} 1`] = `false`;

exports[`matcher search '?foo=bar' not match {"searchParams":"foo="} 1`] = `false`;

exports[`matcher search '?foo=bar' not match {"searchParams":"foo=baz"} 1`] = `false`;
